datasciencebr/jarbas
# Delegate reimbursement import to Celery

This is a work in progress, I just created it earlier to have feedbacks. The aim, when complete, is to address part of #217.

**What is the purpose of this Pull Request?**
The purpose of this PR is make the process of importing reimbursement quicker and without manual steps as described in #217.

**What was done to achieve this purpose?**
I introduced Celery to perform some task asynchronously: that way the heavy work of the import is handled in the background.

**How to test if it really works?**

1. Install RabbitMQ _if not using Docker_ — `brew install rabbitmq` or `apt-get install rabbitmq-server` should do the job for macOS and any Debian based Linux such as Ubuntu (Docker environment is ready, no need to worry about it) 
2. Install new packages: `$ python -m pip install -U -r requirements.txt`
3. Run tests `$ python manage.py test`
4. Get Celery started: `$ celery worker --app jarbas`
5. Run reimbursement import: `$ python manage.py reimbursements <path to reimbursements.xz>`

**Who can help reviewing it?**
@jtemporal @lipemorais @davinirjr @augustogoulart @adrianomargarin

**What is out of scope**
In #217 I mentioned creating a API endpoint so Rosie can send `POST`s with reimbursement data. I think this is an issue for a new PR when this one is finished, reviewed, stable an merged.

**Next steps in this PR**

- [x] Find out why the iteration is taking so long for each CSV (`.xz`) line (either I screwed up when wiring up Celery, or the serialize step is another bottleneck; if so, a callback in Celery should be able to handle serializing and then the create/update step)
- [x] Improve documentation on the RabbitMQ and Celery parts (I added bit on the environment variable, but instructions to run Celery still missing as well as pointing RabbitMQ as a requirement)

**Next steps**

* Bring back Docker Hub images to `docker-compose.yml` after this PR is merged
